sql: handle NOT VISIBLE columns better in pg_catalog

rafiss opened the pull request:
### sql: include hidden columns in pg_index(indkey)

Previously it was impossible to see hidden columns in this table. This was
misleading, since it could make it impossible to inspect columns in an index.

Release note (bug fix): Hidden columns are now included in the pg_catalog.pg_index(indkey) column.

---

### sql: add attishidden column to pg_catalog.pg_attribute

This is useful so that tools that need to inspect columns can avoid
joining to a crdb_internal table to determine if a column is NOT
VISIBLE.

Release note (sql change): The pg_catalog.pg_attribute table now has a
column named attishidden, which indicates if the column is NOT VISIBLE.

---

See context in: https://cockroachlabs.slack.com/archives/C0168LW5THS/p1714514994552219
Epic: None

rafiss committed:
sql: include hidden columns in pg_index(indkey)

Previously it was impossible to see hidden columns in this table. This
was misleading, since it could make it impossible to inspect columns in
an index.

Release note (bug fix): Hidden columns are now included in the
pg_catalog.pg_index(indkey) column.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126397)
<!-- Reviewable:end -->

rafiss committed:
sql: add attishidden column to pg_catalog.pg_attribute

This is useful so that tools that need to inspect columns can avoid
joining to a crdb_internal table to determine if a column is NOT
VISIBLE.

Release note (sql change): The pg_catalog.pg_attribute table now has a
column named attishidden, which indicates if the column is NOT VISIBLE.

fqazi reviewed (approved):
> Reviewed 3 of 3 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126397)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126397#-O0V7djeEhZl7xjZQ9Fp):*
I guess we have done extensions in pg_catalog before to add CRDB specific things before right? Its the last column (relative to PG), so any weird things with SELECT * should be fine.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
tftr!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126397#-:-O0VGm2Y2SPuuVuEp0dr:bjpalj3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @mgartner)

___
*[`-- commits` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126397#-O0V7djeEhZl7xjZQ9Fp):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I guess we have done extensions in pg_catalog before to add CRDB specific things before right? Its the last column (relative to PG), so any weird things with SELECT * should be fine.
</blockquote></details>

yeah, we've added our own columns to the end of the table


<!-- Sent from Reviewable.io -->
